fork_choice/confirmation-rule.md

```diff
@@ -0,0 +1,673 @@
+
+The research paper for this rule is attached in this [ethresear.ch post](https://ethresear.ch/t/confirmation-rule-for-ethereum-pos/15454).
+
+This rule makes the following network synchrony assumption: starting from the current slot, attestations created by honest validators in any slot are received by the end of that slot.
+Consequently, this rule provides confirmations to users who believe in the above assumption. If this assumption is broken, confirmed blocks can be reorged without any adversarial behavior, and without slashing.
+
+There are two algorithms in the document:
+- [**Confirmation Rule**](#confirmation-rule): Given a block and confirmation safety parameters, outputs whether the block is confirmed.
+- [**Confirmation Score**](#confirmation-score): Given a block, outputs the confirmation score for the block, i.e., the maximum possible confirmation safety parameters to deem the block confirmed.
+
+*Note:* These algorithms use unbounded integer arithmetic in some places. The rest of `consensus-specs` uses `uint64` arithmetic exclusively to ensure that results fit into length-limited fields - a property crucial for consensus objects (such as the `BeaconBlockBody`). This document describes a local confirmation rule that does not require storing anything in length-limited fields. Using unbounded integer arithmetic here prevents possible overflowing issues.
+
+## Confirmation Rule
+
+This section specifies an algorithm to determine whether a block is confirmed. The confirmation rule can be configured to the desired tolerance of Byzantine validators, for which the algorithm takes the following input parameters:
+| Input Parameter                    | Type     | Max. Value                         | Description                                                            |
+| ---------------------------------- | -------- |:---------------------------------- | ---------------------------------------------------------------------- |
+| `confirmation_byzantine_threshold` | `uint64` | `33`                               | the maximum percentage of Byzantine validators among the validator set |
+| `confirmation_slashing_threshold`  | `uint64` | `confirmation_byzantine_threshold` | the maximum percentage of slashings among the validator set            |
+
+
+### Helper Functions
+
+#### `get_committee_weight_between_slots`
+
+```python
+def get_committee_weight_between_slots(state: BeaconState, start_slot: Slot, end_slot: Slot) -> Gwei:
+    """Returns the total weight of committees between ``start_slot`` and ``end_slot`` (inclusive of both)..
+    """
+    total_active_balance = get_total_active_balance(state)
+
+    # If an entire epoch is covered by the range, return the total active balance
+    start_epoch = compute_epoch_at_slot(start_slot)
+    end_epoch = compute_epoch_at_slot(end_slot)
+
+    if start_slot > end_slot:
+        return Gwei(0)
```

Comment:
Maybe it would be better to ad an assertion?
```suggestion
    assert start_slot <= end_slot
```

Reply:
This would make the code more complex as `get_remaining_weight_in_current_epoch` may call this method with  `start_slot > end_slot`. Hence, if we add the `assert`, then we need to add the check in `get_remaining_weight_in_current_epoch`.

Reply:
> get_remaining_weight_in_current_epoch logic shouldn't leak into a low level function.

I agree on this.

On the other hand, it does not seem incorrect to me that the committee weight between, say, slot 100 and slot 10 is 0.